man/vlock-main.8
.TH VLOCK-MAIN 8 "28 July 2007" "Linux" "Linux Programmer's Manual"
.SH NAME
vlock-main \- lock current virtual console
.SH SYNOPSIS
.B vlock-main [plugins...]
.SH DESCRIPTION
\fBvlock-main\fR is part of vlock(1), the Virtual Console locking program for
Linux.  It locks the current session and will only exit if the current user can
authenticate themselves.  The root user will also be able to unlock the
session, unless disabled at compile time.  The command line arguments specify
plugins that should be loaded by vlock-main.  See vlock-plugins(5) for more
information.
.PP
If plugin support is disabled at compile time, the only supported argument is
"all".
.SH "ENVIRONMENT VARIABLES"
The following environment variables can be used to change the behavior of
vlock-main:
.PP
.B USER
.IP
If this variable is when vlock-main is run as root (uid 0) vlock locks
the screen as this user instead of root.  The root password will still be able
to unlock the session, unless disabled at compile time.
.PP
.B VLOCK_ALL_MESSAGE
.IP
If this variable is set and all consoles are locked its contents will be used
as the locking message instead of the default message.
.PP
.B VLOCK_CURRENT_MESSAGE
.IP
If this variable is set and only the current consoles is locked its contents
will be used as the locking message instead of the default message.
.PP
.B VLOCK_MESSAGE
.IP
If this variable is set its contents will be used as the locking message
instead of the default.  This overrides the former two variables.
.PP
.B VLOCK_TIMEOUT
.IP
Set this variable to specify the timeout (in seconds) after which the screen
saver plugins (if any) will be invoked.  If this variable is unset or set to an
invalid value or 0 no timeout is used.  See vlock-plugins(5) for more
information about plugins.
.PP
.B VLOCK_PROMPT_TIMEOUT
.IP
Set this variable to specify the amount of time (in seconds) you will have to
enter your password at the password prompt.  If this variable is unset or set
to an invalid value or 0 no timeout is used.  \fBWarning\fR: If this value is
too low, you may not be able to unlock your session.
.PP
.SH SIGNALS
Several signals are ignored.  \fBvlock-main\fR will try to exit cleanly if
SIGTERM is received.
.SH "SEE ALSO"
.BR vlock (1),
.BR vlock-plugins (5)
.SH AUTHORS
Frank Benkstein <frank-vlock@benkstein.net>


ChangeLog
2011-09-18 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2.3

	* Fix vlock not reacting to input when started in the background
	from bash.  (Thanks to Amy Fong for the report.)

2008-05-14 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2.2

	* Fix and improve build system: use LDLIBS instead of LDFLAGS where
	appropriate and provide EXTRA_CFLAGS/EXTRA_LDFLAGS parameters in ./configure.
	* Display number of authentication tries on exit.

2008-03-26 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2.1

	* Fix a script handling bug.

2008-03-21 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2

	* Small plugin documentation update.

2008-01-15 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc5

	* Fixed snprintf() corner case in "new" module.
	* Restore timeout if select() was interrupted by a signal in prompt().

2007-12-29 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc4

	* Slightly changed the fix for the previous security issue: directory parts
	are now silently stripped from erroneous names instead of reported as errors.
	* Fixed a file descriptor leak in script handling.
	* Fixed a bug where vlock would run in an infinite loop if not started from
	a terminal.  Now it simply exits.
	* Added unit tests (using cunit).

2007-11-28 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc3

	* A critical security problem (local root exploit) was fixed:
		Previous versions of vlock (since 2.2 alpha1) contained a serious flaw
		that allowed any user to execute arbitrary code as root .  This was
		possible because plugin names could contain "../" thus escaping the
		pre-defined plugin directory.  All users of vlock 2.2 development versions
		are advised to update.
	* Improved script handling.
	* Added two more scripts.

2007-11-26 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc2

	* Fixed man page installation error.

2007-11-17 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc1

	* Updated documentation.

2007-11-04 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 beta2

	* Improved error handling and error messages.
	* Fixed portability bugs.
	* Fixed dependency handling bug.

2007-10-16 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 beta1

	* Documentation update.
	* Improved error handling.

2007-10-03 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 alpha3

	* Added ./configure script.
	* Improved plugin handling.
	* Improved error handling.
	* Improved child process handling.

2007-09-21 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 alpha2

	* Added libcaca based screen saver module.
	* Miscellaneous minor fixes.

2007-09-19 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 alpha1

	* Medium rewrite was undertaken.
	* Support for plugins (modules and scripts) was added.
	* vlock-all, vlock-new, and vlock-nosysrq were converted to modules.
	* Other sample modules and scripts were added.
	* vlock now tries to terminale cleanly if killed by SIGTERM and on errors.

2007-09-08 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1

	* Documentation update.

2007-08-29 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1 rc1

	* Really tiny cleanups.

2007-08-23 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1 beta1

	* Fix compilation on Debian/kFreeBSD.

2007-08-16 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1 alpha2

	* Locking message now configurable.
	* Added configuration file support: ~/.vlockrc.
	* "vlock --new" now works even if stdin is not a terminal.
	* Improved documentation.

2007-08-11 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1 alpha1

	* Added FreeBSD support.  Build with PAM_LIBS=-lpam, install with
	  INSTALL=ginstall.
	* Improved signal handling.
	* Added timeout support to password prompts.

2007-08-09 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0

	* No changes since vlock 2.0 rc2.

2007-08-06 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 rc2

	* Improved man page and error messages.
	* Fixed bug that prevented shadow authentication from working.
	* vlock can now only be killed by root.  This will probably be
	  fixed in future releases.

2007-08-05 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 rc1

	* Code cleanups (comments, style).
	* Fixed two merely theoretical security problems.

2007-08-02 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 beta2

	* Many improvements to security.
	* General cleanup to source tree.
	* ChangeLog added.
	* Manpages updated.
	* LICENSE file updated, license terms unchanged.
	* README updated.
	* Added a SECURITY blurb.
	* -n,--new option now implies -a,--all instead of requiring it.

2007-07-30 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 beta1

	* Added manpages for all tools.
	* Enter key must be pressed before authentication is started.
	* When run as root, vlock now locks as the user given by the
	  $USER environment variable.
	* Add -n,--new option to run vlock on a new virtual console.

2007-07-28 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 alpha2

	* Shadow authentication added.

2007-07-25 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 alpha1

	* Complete rewrite. vlock is now made of several tools called
	  from a shell script.

2007-05-23 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 1.4

	* No changes since rc2.

2007-05-10 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 1.4 rc2

	* Make install rule made more friendly to packagers.
	* Updated license declarations: sources now clearly
	  state "GNU General Public License version 2".

2007-05-09 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 1.4 rc1

	* Some compile time warnings fixed..
	* Support for SysRq disabling added.

1999-01-13 Michael Johnson <johnsonm@redhat.com>

	vlock 1.3

1998-03-12 Michael Johnson <johnsonm@redhat.com>

	vlock 1.2


src/vlock-main.c
/* vlock-main.c -- main routine for vlock,
 *                    the VT locking program for linux
 *
 * This program is copyright (C) 2007 Frank Benkstein, and is free
 * software which is freely distributable under the terms of the
 * GNU General Public License version 2, included as the file COPYING in this
 * distribution.  It is NOT public domain software, and any
 * redistribution not permitted by the GNU General Public License is
 * expressly forbidden without prior written permission from
 * the author.
 *
 */

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include <pwd.h>

#include <termios.h>
#include <unistd.h>
#include <sys/types.h>
#include <signal.h>
#include <errno.h>
#include <time.h>

#include "prompt.h"
#include "auth.h"
#include "console_switch.h"
#include "util.h"

#ifdef USE_PLUGINS
#include "plugins.h"
#endif

int vlock_debug = 0;

#define ensure_atexit(func) \
  do { \
    if (atexit(func) != 0) \
      fatal_perror("vlock: atexit() failed"); \
  } while (0)

static char *get_username(void)
{
  uid_t uid = getuid();
  char *username = NULL;

  /* Get the user name from the environment if started as root. */
  if (uid == 0)
    username = getenv("USER");

  if (username == NULL) {
    struct passwd *pw;

    /* Get the password entry. */
    pw = getpwuid(uid);

    if (pw == NULL)
      return NULL;

    username = pw->pw_name;
  }

  return strdup(username);
}

static void terminate(int signum)
{
  fprintf(stderr, "vlock: Terminated!\n");
  /* Call exit here to ensure atexit handlers are called. */
  exit(1);
}

static void block_signals(void)
{
  struct sigaction sa;

  /* Ignore some signals. */
  /* These signals shouldn't be delivered anyway, because terminal signals are
   * disabled below. */
  (void) sigemptyset(&(sa.sa_mask));
  sa.sa_flags = SA_RESTART;
  sa.sa_handler = SIG_IGN;
  (void) sigaction(SIGINT, &sa, NULL);
  (void) sigaction(SIGQUIT, &sa, NULL);
  (void) sigaction(SIGTSTP, &sa, NULL);

  /* Install special handler for SIGTERM. */
  sa.sa_flags = SA_RESETHAND;
  sa.sa_handler = terminate;
  (void) sigaction(SIGTERM, &sa, NULL);
}

static struct termios old_term;

static void setup_terminal(void)
{
  (void) tcgetattr(STDIN_FILENO, &old_term);
  struct termios new_term = old_term;
  /* Pressing enter must yield line feed. */
  new_term.c_iflag &= ~INLCR;
  new_term.c_iflag |= ICRNL;
  /* Disable terminal echoing and signals. */
  new_term.c_lflag &= ~(ECHO | ISIG);
  (void) tcsetattr(STDIN_FILENO, TCSANOW, &new_term);
}

static void restore_terminal(void)
{
  /* Restore the terminal. */
  (void) tcsetattr(STDIN_FILENO, TCSANOW, &old_term);
}

static int auth_tries;

static void auth_loop(const char *username)
{
  struct timespec *prompt_timeout;
  struct timespec *wait_timeout;
  char *vlock_message;

  /* Get the vlock message from the environment. */
  vlock_message = getenv("VLOCK_MESSAGE");

  if (vlock_message == NULL) {
    if (console_switch_locked)
      vlock_message = getenv("VLOCK_ALL_MESSAGE");
    else
      vlock_message = getenv("VLOCK_CURRENT_MESSAGE");
  }

  /* Get the timeouts from the environment. */
  prompt_timeout = parse_seconds(getenv("VLOCK_PROMPT_TIMEOUT"));
#ifdef USE_PLUGINS
  wait_timeout = parse_seconds(getenv("VLOCK_TIMEOUT"));
#else
  wait_timeout = NULL;
#endif

  for (;;) {
    char c;

    /* Print vlock message if there is one. */
    if (vlock_message && *vlock_message) {
      fputs(vlock_message, stderr);
      fputc('\n', stderr);
    }

    /* Wait for enter or escape to be pressed. */
    c = wait_for_character("\n\033", wait_timeout);

    /* Escape was pressed or the timeout occurred. */
    if (c == '\033' || c == 0) {
#ifdef USE_PLUGINS
      plugin_hook("vlock_save");
      /* Wait for any key to be pressed. */
      c = wait_for_character(NULL, NULL);
      plugin_hook("vlock_save_abort");

      /* Do not require enter to be pressed twice. */
      if (c != '\n')
        continue;
#else
      continue;
#endif
    }

    /* Try authentication as user. */
    if (auth(username, prompt_timeout))
      break;
    else
      sleep(1);

#ifndef NO_ROOT_PASS
    if (strcmp(username, "root") != 0) {
      /* Try authentication as root. */
      if (auth("root", prompt_timeout))
        break;
      else
        sleep(1);
    }
#endif

    auth_tries++;
  }

  /* Free timeouts memory. */
  free(wait_timeout);
  free(prompt_timeout);
}

void display_auth_tries(void)
{
  if (auth_tries > 0)
    fprintf(stderr, "%d failed authentication %s.\n", auth_tries, auth_tries > 1 ? "tries" : "try");
}

#ifdef USE_PLUGINS
static void call_end_hook(void)
{
  (void) plugin_hook("vlock_end");
}
#endif

/* Lock the current terminal until proper authentication is received. */
int main(int argc, char *const argv[])
{
  char *username;

  vlock_debug = (getenv("VLOCK_DEBUG") != NULL);

  block_signals();

  username = get_username();

  if (username == NULL)
    fatal_perror("vlock: could not get username");

  ensure_atexit(display_auth_tries);

#ifdef USE_PLUGINS
  for (int i = 1; i < argc; i++)
    if (!load_plugin(argv[i]))
      fatal_error("vlock: loading plugin '%s' failed: %s", argv[i], STRERROR);

  ensure_atexit(unload_plugins);

  if (!resolve_dependencies()) {
    if (errno == 0)
      exit(EXIT_FAILURE);
    else
      fatal_error("vlock: error resolving plugin dependencies: %s", STRERROR);
  }

  plugin_hook("vlock_start");
  ensure_atexit(call_end_hook);
#else /* !USE_PLUGINS */
  /* Emulate pseudo plugin "all". */
  if (argc == 2 && (strcmp(argv[1], "all") == 0)) {
    if (!lock_console_switch()) {
      if (errno)
        perror("vlock: could not disable console switching");

      exit(EXIT_FAILURE);
    }

    ensure_atexit((void (*)(void))unlock_console_switch);
  } else if (argc > 1) {
    fatal_error("vlock: plugin support disabled");
  }
#endif

  if (!isatty(STDIN_FILENO))
    fatal_error("vlock: stdin is not a terminal");

  setup_terminal();
  ensure_atexit(restore_terminal);

  auth_loop(username);

  free(username);

  exit(0);
}
